File: shared/engine/elo.js
// ═══════════════════════════════════════════════════════════════════
//  Nova Edge — موتور ELO / امتیاز رنک (RP) / سطح مهارت
//  ⚠️ این فایل «منبع حقیقت» ریاضیات رتبه‌بندی است.
//     پورت دقیق آن در:
//       java-plugin/.../core/EloEngine.java
//       bedrock-plugin/src/NovaEdge/core/EloEngine.php
//     وجود دارد و هر سه با shared/testvectors.json تست می‌شوند.
// ═══════════════════════════════════════════════════════════════════
import { clamp } from './rng.js';

export const ELO = {
  FLOOR: 100,
  CAP: 4000,
  START: 1000,
  PROVISIONAL_GAMES: 10,
  PROVISIONAL_K_MULT: 1.5,
  HIGH_RATING: 2200,
  HIGH_K_MULT: 0.75,
  VERY_HIGH_RATING: 2600,
  VERY_HIGH_K_MULT: 0.6,
  DECAY_AFTER_DAYS: 21,
  DECAY_PER_WEEK: 8,
  DECAY_MIN_RATING: 1400,
};

export function expectedScore(ratingA, ratingB) {
  const a = Number(ratingA) || ELO.START;
  const b = Number(ratingB) || ELO.START;
  return 1 / (1 + Math.pow(10, (b - a) / 400));
}

/** میانگین رتبهٔ تیم (برای مودهای تیمی) */
export function teamRating(ratings) {
  const list = (ratings || []).map((r) => Number(r) || ELO.START).filter((n) => Number.isFinite(n));
  if (!list.length) return ELO.START;
  return list.reduce((s, n) => s + n, 0) / list.length;
}

/**
 * ضریب K مؤثر: مود + وضعیت پروویژنال + رتبهٔ بالا
 * @returns {number}
 */
export function effectiveK(baseK, gamesPlayed, rating) {
  let k = Math.max(4, Number(baseK) || 24);
  if ((Number(gamesPlayed) || 0) < ELO.PROVISIONAL_GAMES) k *= ELO.PROVISIONAL_K_MULT;
  if ((Number(rating) || 0) >= ELO.VERY_HIGH_RATING) k *= ELO.VERY_HIGH_K_MULT;
  else if ((Number(rating) || 0) >= ELO.HIGH_RATING) k *= ELO.HIGH_K_MULT;
  return Math.round(k * 100) / 100;
}

/**
 * امتیاز واقعی S از روی جایگاه در مودهای FFA
 * نفر اول = ۱، نفر آخر = ۰
 */
export function placementScore(place, playerCount) {
  const p = clamp(Number(place) || 1, 1, Math.max(1, Number(playerCount) || 1));
  const n = Math.max(2, Number(playerCount) || 2);
  return (n - p) / (n - 1);
}

/**
 * محاسبهٔ تغییر ELO برای یک بازیکن/تیم
 * @param {object} o {rating, opponentRating, baseK, gamesPlayed, score}
 * @returns {{delta:number, expected:number, k:number, rating:number}}
 */
export function eloDelta(o = {}) {
  const rating = clamp(Number(o.rating) || ELO.START, ELO.FLOOR, ELO.CAP);
  const opp = clamp(Number(o.opponentRating) || ELO.START, ELO.FLOOR, ELO.CAP);
  const e = expectedScore(rating, opp);
  const s = clamp(Number(o.score) || 0, 0, 1);
  const k = effectiveK(o.baseK, o.gamesPlayed, rating);
  const delta = Math.round(k * (s - e));
  return { delta, expected: Math.round(e * 10000) / 10000, k, rating: clamp(rating + delta, ELO.FLOOR, ELO.CAP) };
}

/**
 * نتایج کامل یک مچ برای همهٔ بازیکنان (انسان و بات)
 * @param {object} o {mode, players, placements, teams}
 *  teams: [{index, players:[{id}]}] — اگر داده شود، رتبهٔ حریف از تیم مقابل
 *  گرفته می‌شود (دقیق‌تر از میانگین کل برای مودهای تیمی).
 */
export function settleMatch({ mode, players = [], placements = {}, teams = null }) {
  const baseK = Number(mode?.match?.elo_k) || 24;
  const category = String(mode?.category || 'ffa');
  const humans = players.filter((p) => !p.is_bot);
  const n = players.length;

  // نگاشت بازیکن → تیم
  const teamOf = new Map();
  const teamRatingOf = new Map();
  if (Array.isArray(teams) && teams.length > 1) {
    for (const t of teams) {
      const ids = (t.players || []).map((p) => String(p.id ?? p));
      const list = players.filter((p) => ids.includes(String(p.id)));
      for (const id of ids) teamOf.set(id, t.index);
      teamRatingOf.set(t.index, teamRating(list.map((p) => Number(p.rating) || ELO.START)));
    }
  }

  const total = players.reduce((s, p) => s + clamp(Number(p.rating) || ELO.START, ELO.FLOOR, ELO.CAP), 0);
  const out = [];
  for (const p of players) {
    const my = clamp(Number(p.rating) || ELO.START, ELO.FLOOR, ELO.CAP);
    let oppRating;
    const myTeam = teamOf.get(String(p.id));
    if (myTeam !== undefined && teamRatingOf.size > 1) {
      // میانگین رتبهٔ تیم‌های مقابل
      const others = [...teamRatingOf.entries()].filter(([idx]) => idx !== myTeam).map(([, r]) => r);
      oppRating = others.length ? others.reduce((s, r) => s + r, 0) / others.length : my;
    } else {
      oppRating = n > 1 ? (total - my) / (n - 1) : my;
    }
    let score;
    if (category === 'coop' || category === 'social') {
      score = Number(p.won) ? 1 : 0;
    } else if (category === 'ffa' || category === 'race' || category === 'creative') {
      const place = Number(placements[p.id] ?? p.place ?? (p.won ? 1 : n));
      score = placementScore(place, n);
    } else {
      score = Number(p.won) ? 1 : 0;
    }
    const r = eloDelta({ rating: my, opponentRating: oppRating, baseK, gamesPlayed: p.games_played ?? p.games ?? 0, score });
    out.push({
      id: p.id,
      is_bot: !!p.is_bot,
      before: my,
      after: r.rating,
      delta: r.delta,
      expected: r.expected,
      score,
      rp: p.is_bot ? 0 : rankPointsFrom({ score_points: Number(p.score_points) || 0, elo_delta: r.delta }),
    });
  }
  return { category, base_k: baseK, human_count: humans.length, results: out };
}

/** امتیاز رنک (RP) از روی امتیاز مچ + تغییر ELO */
export function rankPointsFrom({ score_points = 0, elo_delta = 0 } = {}) {
  const base = Math.floor(Math.max(0, Number(score_points) || 0) / 10);
  const bonus = Math.floor(Math.max(0, Number(elo_delta) || 0) / 4);
  return Math.max(0, base + bonus);
}

/** رنک بر اساس RP (فقط نردبان امتیازی، بدون خرید) */
export function rankForRp(ranks, rp) {
  const list = (ranks || []).slice().sort((a, b) => (a.rp_required || 0) - (b.rp_required || 0));
  let best = list[0] || { id: 'free', index: 0 };
  for (const r of list) if ((Number(r.rp_required) || 0) <= (Number(rp) || 0)) best = r;
  return best;
}

/** رنک نهایی = بیشینهٔ (کسب‌شده با بازی، خریداری‌شده) */
export function effectiveRank(ranks, rp, purchasedRankId) {
  const list = (ranks || []).slice().sort((a, b) => (a.index || 0) - (b.index || 0));
  const earned = rankForRp(list, rp);
  const bought = list.find((r) => r.id === purchasedRankId) || null;
  if (bought && (bought.index || 0) > (earned.index || 0)) return { rank: bought, source: 'purchased' };
  return { rank: earned, source: bought ? 'earned_over_purchased' : 'earned' };
}

/** پیشرفت تا رنک بعدی (برای نوار پیشرفت UI) */
export function rankProgress(ranks, rp) {
  const list = (ranks || []).slice().sort((a, b) => (a.rp_required || 0) - (b.rp_required || 0));
  const cur = rankForRp(list, rp);
  const next = list.find((r) => (r.rp_required || 0) > (cur.rp_required || 0));
  if (!next) return { current: cur.id, next: null, pct: 100, remaining_rp: 0 };
  const span = (next.rp_required || 0) - (cur.rp_required || 0);
  const done = (Number(rp) || 0) - (cur.rp_required || 0);
  return {
    current: cur.id,
    next: next.id,
    pct: Math.round(clamp((done / Math.max(1, span)) * 100, 0, 100)),
    remaining_rp: Math.max(0, Math.ceil((next.rp_required || 0) - (Number(rp) || 0))),
  };
}

/** سطح (Level) از XP — فرمول economy.json */
export function levelForXp(xp) {
  const x = Math.max(0, Number(xp) || 0);
  return Math.min(200, Math.floor(Math.sqrt(x / 100)));
}
export function xpForLevel(level) {
  const l = Math.max(0, Number(level) || 0);
  return l * l * 100;
}

/**
 * شاخص مهارت ۰..۱ — ورودی اصلی انتخاب سطح هوش بات‌ها.
 * ترکیب ELO (وزن اصلی)، لول، رنک و سابقهٔ بازی.
 */
export function skillIndex({ rating = ELO.START, level = 1, rank_index = 0, games = 0, bias = 0 } = {}) {
  const eloPart = clamp(((Number(rating) || ELO.START) - 700) / 2000, 0, 1); // 700→0 , 2700→1
  const levelPart = clamp((Number(level) || 1) / 120, 0, 1);
  const rankPart = clamp((Number(rank_index) || 0) / 5, 0, 1);
  const expPart = clamp((Number(games) || 0) / 400, 0, 1);
  const raw = 0.5 * eloPart + 0.2 * levelPart + 0.2 * rankPart + 0.1 * expPart;
  return Math.round(clamp(raw + (Number(bias) || 0), 0, 1) * 1000) / 1000;
}

/** افول رتبه برای بازیکنان غیرفعال (فقط بالای آستانه) */
export function decayRating(rating, daysInactive) {
  const r = Number(rating) || ELO.START;
  if (r <= ELO.DECAY_MIN_RATING) return r;
  const weeks = Math.max(0, Math.floor(((Number(daysInactive) || 0) - ELO.DECAY_AFTER_DAYS) / 7));
  if (weeks <= 0) return r;
  return Math.max(ELO.DECAY_MIN_RATING, r - weeks * ELO.DECAY_PER_WEEK);
}

File: server/java/src/main/java/ir/novaedge/core/Elo.java
// ═══════════════════════════════════════════════════════════════════
//  Nova Edge — core/Elo.java (پورت دقیق shared/engine/elo.js)
// ═══════════════════════════════════════════════════════════════════
package ir.novaedge.core;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public final class Elo {

    public static final double FLOOR = 100;
    public static final double CAP = 4000;
    public static final double START = 1000;
    public static final int PROVISIONAL_GAMES = 10;
    public static final double PROVISIONAL_K_MULT = 1.5;
    public static final double HIGH_RATING = 2200;
    public static final double HIGH_K_MULT = 0.75;
    public static final double VERY_HIGH_RATING = 2600;
    public static final double VERY_HIGH_K_MULT = 0.6;
    public static final int DECAY_AFTER_DAYS = 21;
    public static final double DECAY_PER_WEEK = 8;
    public static final double DECAY_MIN_RATING = 1400;

    private Elo() {
    }

    public static double expectedScore(double a0, double b0) {
        double a = a0 != 0 ? a0 : START;
        double b = b0 != 0 ? b0 : START;
        return 1 / (1 + Math.pow(10, (b - a) / 400));
    }

    public static double teamRating(List<Double> ratings) {
        List<Double> list = new ArrayList<>();
        for (Double r : ratings) {
            double n = r != null && r != 0 ? r : START;
            if (Double.isFinite(n)) {
                list.add(n);
            }
        }
        if (list.isEmpty()) {
            return START;
        }
        double s = 0;
        for (double n : list) {
            s += n;
        }
        return s / list.size();
    }

    public static double effectiveK(double baseK, int gamesPlayed, double rating) {
        double k = Math.max(4, baseK != 0 ? baseK : 24);
        if (gamesPlayed < PROVISIONAL_GAMES) {
            k *= PROVISIONAL_K_MULT;
        }
        if (rating >= VERY_HIGH_RATING) {
            k *= VERY_HIGH_K_MULT;
        } else if (rating >= HIGH_RATING) {
            k *= HIGH_K_MULT;
        }
        return Rng.roundTo(k, 2);
    }

    public static double placementScore(double place, int playerCount) {
        double p = Rng.clamp(place != 0 ? place : 1, 1, Math.max(1, playerCount));
        double n = Math.max(2, playerCount);
        return (n - p) / (n - 1);
    }

    public static Map<String, Object> eloDelta(Map<String, Object> o) {
        double rating = Rng.clamp(Json.numOr(o, "rating", START), FLOOR, CAP);
        double opp = Rng.clamp(Json.numOr(o, "opponentRating", START), FLOOR, CAP);
        double e = expectedScore(rating, opp);
        double s = Rng.clamp(Json.num(o, "score", 0), 0, 1);
        double k = effectiveK(Json.num(o, "baseK", 24), (int) Json.num(o, "gamesPlayed", 0), rating);
        long delta = Rng.jsRound(k * (s - e));
        Map<String, Object> out = Json.obj();
        Json.put(out, "delta", delta);
        Json.put(out, "expected", Rng.roundTo(e, 4));
        Json.put(out, "k", k);
        Json.put(out, "rating", Rng.clamp(rating + delta, FLOOR, CAP));
        return out;
    }

    public static Map<String, Object> settleMatch(Map<String, Object> mode, List<Map<String, Object>> players,
            Map<String, Object> placements, List<Map<String, Object>> teams) {
        double baseK = Json.numOr(Json.map(mode, "match"), "elo_k", 24);
        String category = Json.str(mode, "category", "ffa");
        int humans = 0;
        for (Map<String, Object> p : players) {
            if (!Json.bool(p, "is_bot", false)) {
                humans++;
            }
        }
        int n = players.size();

        Map<String, Integer> teamOf = new HashMap<>();
        Map<Integer, Double> teamRatingOf = new HashMap<>();
        if (teams != null && teams.size() > 1) {
            for (Map<String, Object> t : teams) {
                List<String> ids = new ArrayList<>();
                for (Map<String, Object> pp : Json.mapList(t.get("players"))) {
                    ids.add(Json.str(pp, "id", String.valueOf(pp.get("id"))));
                }
                List<Double> rlist = new ArrayList<>();
                for (Map<String, Object> p : players) {
                    if (ids.contains(Json.str(p, "id", ""))) {
                        rlist.add(Json.numOr(p, "rating", START));
                    }
                }
                for (String id : ids) {
                    teamOf.put(id, (int) Json.num(t, "index", 0));
                }
                teamRatingOf.put((int) Json.num(t, "index", 0), teamRating(rlist));
            }
        }

        double total = 0;
        for (Map<String, Object> p : players) {
            total += Rng.clamp(Json.numOr(p, "rating", START), FLOOR, CAP);
        }

        List<Map<String, Object>> results = new ArrayList<>();
        for (Map<String, Object> p : players) {
            double my = Rng.clamp(Json.numOr(p, "rating", START), FLOOR, CAP);
            Integer myTeam = teamOf.get(Json.str(p, "id", ""));
            double oppRating;
            if (myTeam != null && teamRatingOf.size() > 1) {
                List<Double> others = new ArrayList<>();
                for (Map.Entry<Integer, Double> en : teamRatingOf.entrySet()) {
                    if (!en.getKey().equals(myTeam)) {
                        others.add(en.getValue());
                    }
                }
                double sum = 0;
                for (double d : others) {
                    sum += d;
                }
                oppRating = others.isEmpty() ? my : sum / others.size();
            } else {
                oppRating = n > 1 ? (total - my) / (n - 1) : my;
            }
            double score;
            if (category.equals("coop") || category.equals("social")) {
                score = Json.bool(p, "won", false) ? 1 : 0;
            } else if (category.equals("ffa") || category.equals("race") || category.equals("creative")) {
                Object place = placements != null ? placements.get(p.get("id")) : null;
                double pl = place != null ? Json.num(place, 0) : (p.get("place") != null ? Json.num(p, "place", 0) : (Json.bool(p, "won", false) ? 1 : n));
                score = placementScore(pl, n);
            } else {
                score = Json.bool(p, "won", false) ? 1 : 0;
            }
            Map<String, Object> arg = Json.obj();
            Json.put(arg, "rating", my);
            Json.put(arg, "opponentRating", oppRating);
            Json.put(arg, "baseK", baseK);
            Json.put(arg, "gamesPlayed", (int) Json.num(p, "games_played", Json.num(p, "games", 0)));
            Json.put(arg, "score", score);
            Map<String, Object> r = eloDelta(arg);
            Map<String, Object> row = Json.obj();
            Json.put(row, "id", p.get("id"));
            Json.put(row, "is_bot", Json.bool(p, "is_bot", false));
            Json.put(row, "before", my);
            Json.put(row, "after", r.get("rating"));
            Json.put(row, "delta", r.get("delta"));
            Json.put(row, "expected", r.get("expected"));
            Json.put(row, "score", score);
            Json.put(row, "rp", Json.bool(p, "is_bot", false) ? 0
                    : rankPointsFrom(Json.num(p, "score_points", 0), ((Number) r.get("delta")).doubleValue()));
            results.add(row);
        }
        Map<String, Object> out = Json.obj();
        Json.put(out, "category", category);
        Json.put(out, "base_k", baseK);
        Json.put(out, "human_count", humans);
        Json.put(out, "results", results);
        return out;
    }

    public static long rankPointsFrom(double scorePoints, double eloDelta) {
        long base = (long) Math.floor(Math.max(0, scorePoints) / 10);
        long bonus = (long) Math.floor(Math.max(0, eloDelta) / 4);
        return Math.max(0, base + bonus);
    }

    public static Map<String, Object> rankForRp(List<Map<String, Object>> ranks, double rp) {
        List<Map<String, Object>> list = new ArrayList<>(ranks);
        list.sort(Comparator.comparingDouble(m -> Json.num(m, "rp_required", 0)));
        Map<String, Object> best = list.isEmpty() ? Json.obj() : list.get(0);
        for (Map<String, Object> r : list) {
            if (Json.num(r, "rp_required", 0) <= rp) {
                best = r;
            }
        }
        return best;
    }

    public static Map<String, Object> effectiveRank(List<Map<String, Object>> ranks, double rp, String purchasedRankId) {
        List<Map<String, Object>> list = new ArrayList<>(ranks);
        list.sort(Comparator.comparingDouble(m -> Json.num(m, "index", 0)));
        Map<String, Object> earned = rankForRp(list, rp);
        Map<String, Object> bought = null;
        for (Map<String, Object> r : list) {
            if (Json.str(r, "id", "").equals(purchasedRankId)) {
                bought = r;
                break;
            }
        }
        Map<String, Object> out = Json.obj();
        if (bought != null && Json.num(bought, "index", 0) > Json.num(earned, "index", 0)) {
            Json.put(out, "rank", bought);
            Json.put(out, "source", "purchased");
        } else {
            Json.put(out, "rank", earned);
            Json.put(out, "source", bought != null ? "earned_over_purchased" : "earned");
        }
        return out;
    }

    public static Map<String, Object> rankProgress(List<Map<String, Object>> ranks, double rp) {
        List<Map<String, Object>> list = new ArrayList<>(ranks);
        list.sort(Comparator.comparingDouble(m -> Json.num(m, "rp_required", 0)));
        Map<String, Object> cur = rankForRp(list, rp);
        Map<String, Object> next = null;
        for (Map<String, Object> r : list) {
            if (Json.num(r, "rp_required", 0) > Json.num(cur, "rp_required", 0)) {
                next = r;
                break;
            }
        }
        Map<String, Object> out = Json.obj();
        if (next == null) {
            Json.put(out, "current", Json.str(cur, "id", ""));
            Json.put(out, "next", null);
            Json.put(out, "pct", 100);
            Json.put(out, "remaining_rp", 0);
            return out;
        }
        double span = Json.num(next, "rp_required", 0) - Json.num(cur, "rp_required", 0);
        double done = rp - Json.num(cur, "rp_required", 0);
        Json.put(out, "current", Json.str(cur, "id", ""));
        Json.put(out, "next", Json.str(next, "id", ""));
        Json.put(out, "pct", Rng.jsRound(Rng.clamp(done / Math.max(1, span) * 100, 0, 100)));
        Json.put(out, "remaining_rp", (long) Math.max(0, Math.ceil(Json.num(next, "rp_required", 0) - rp)));
        return out;
    }

    public static long levelForXp(double xp) {
        return Math.min(200, (long) Math.floor(Math.sqrt(Math.max(0, xp) / 100)));
    }

    public static long xpForLevel(long level) {
        long l = Math.max(0, level);
        return l * l * 100;
    }

    public static double skillIndex(Map<String, Object> o) {
        double rating = Json.numOr(o, "rating", START);
        double eloPart = Rng.clamp((rating - 700) / 2000, 0, 1);
        double levelPart = Rng.clamp(Json.num(o, "level", 1) / 120, 0, 1);
        double rankPart = Rng.clamp(Json.num(o, "rank_index", 0) / 5, 0, 1);
        double expPart = Rng.clamp(Json.num(o, "games", 0) / 400, 0, 1);
        double raw = 0.5 * eloPart + 0.2 * levelPart + 0.2 * rankPart + 0.1 * expPart;
        return Rng.roundTo(Rng.clamp(raw + Json.num(o, "bias", 0), 0, 1), 3);
    }

    public static double decayRating(double rating, int daysInactive) {
        double r = rating != 0 ? rating : START;
        if (r <= DECAY_MIN_RATING) {
            return r;
        }
        long weeks = Math.max(0, (long) Math.floor((daysInactive - DECAY_AFTER_DAYS) / 7.0));
        if (weeks <= 0) {
            return r;
        }
        return Math.max(DECAY_MIN_RATING, r - weeks * DECAY_PER_WEEK);
    }
}
